src/jit/optcse.cpp

```diff
@@ -2574,11 +2574,29 @@ bool Compiler::optIsCSEcandidate(GenTree* tree)
         return false;
     }
 
-    // If this is a struct type, we can only consider it for CSE-ing if we can get at
-    // its handle, so that we can create a temp.
-    if ((type == TYP_STRUCT) && (gtGetStructHandleIfPresent(tree) == NO_CLASS_HANDLE))
+    // For some struct types, we now allow them to be CSE candidates.
+    //
+    if (type == TYP_STRUCT)
     {
-        return false;
+        // We can only consider it for CSE-ing if we can get at
+        // its handle, so that we can create a temp.
+        //
+        CORINFO_CLASS_HANDLE structHnd = gtGetStructHandleIfPresent(tree);
+        if (structHnd == NO_CLASS_HANDLE)
+        {
+            return false;
+        }
+
+        unsigned          size = info.compCompHnd->getClassSize(structHnd);
+        structPassingKind howToPassStruct;
+        var_types         structBaseType = getArgTypeForStruct(structHnd, &howToPassStruct, false, size);
+
+        // Only allow structs whose storage can be held in a register to be CSE candidates.
```

Comment:
Is "register" referring just to integer registers and to the lower element of vector registers (for float/double)?

I imagine we may want to support CSE (and other optimizations) for full vector registers eventually...

Reply:
The call to `getArgTypeForStruct`  should identify the register sized SIMD struct types as 
` (howToPassStruct ==  SPK_PrimitiveType)`

Reply:
Currently it doesn't appear that we support passing such vector types in registers  according to `getArgTypeForStruct`

Reply:
I believe @CarolEidt is working on this for the Vector ABI support.